### PR TITLE
Introduce the !showflags jdmpview command

### PR DESCRIPTION
**Pull request by dusanboskovic**
showflags will output the user's specified flag structure and the corresponding values of each flag.

There are 3 method of running this command.
Method 1: !showflags
When running just the !showflags command, it will output all of the possible flag structures that are available

Method 2: !showflags flag_structure
When specifying a certain flag structure, the constants of that flag structure will be outputted (along with all of their values)

Method 3: !showflags flag_structure specific_flag
When specifiying a certain flag, along with the structure that it belongs to, the corresponding value will be outputted

Signed-off-by: Dusan-Boskovic <dusan.boskovic@ibm.com>

### CODE REVIEW

**Comment by dusanboskovic** [comment]
This command in is preparation of a DDR test that will verify if the a system core file was created by an OpenJ9 build or an IBM Java build

Reviewer: @keithc-ca

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
+/*******************************************************************************
+ * Copyright (c) 2001, 2020 IBM Corp. and others
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:2]: The first year should be 2020.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+package com.ibm.j9ddr.vm29.tools.ddrinteractive.commands;
+
+import java.io.PrintStream;
+import java.util.Collection;
+
+import com.ibm.j9ddr.tools.ddrinteractive.Context;
+import com.ibm.j9ddr.tools.ddrinteractive.DDRInteractiveCommandException;
+import com.ibm.j9ddr.tools.ddrinteractive.Command;
+import com.ibm.j9ddr.StructureReader.StructureDescriptor;
+import com.ibm.j9ddr.StructureReader.ConstantDescriptor;
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:31]: Please organize imports (sort by name).

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+
+
+
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:35]: Please collapse multiple blank lines to a single blank line.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+public class ShowFlags extends Command {
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:36]: This should be `ShowFlagsCommand` following the established convention.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    
+    public ShowFlags() {
+		addCommand("showflags", "[flags structure] [specific flag]" , "Retreive the specified flags and their value");
+    }
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:40]: Please indent consistently (using tabs like line 39).

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            return;
+        }
+
+        StructureDescriptor currentStruct = null;
+        ConstantDescriptor currentFlag = null;
+        Collection<StructureDescriptor> currentStructures = context.vmData.getStructures();
+
+        //If no arguments are specified, output all the structures
+        if (args.length == 0) {
+            out.println("Possible flags: ");
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:54]: Please remove trailing space in output.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                }
+            }
+        }
+
+        //Output all the constants of the flags structure 
+        if (args.length > 0){
+            for (StructureDescriptor i : currentStructures){
+                if (i.getName().equals(args[0])){
+                    currentStruct = i;
+                    if (args.length == 2){
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:67]: Please format consistently throughout (e.g. space before `{`).

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                        }
+                    }
+                }
+            }
+            if (currentStruct == null){ 
+                out.println("That structure does not exist");
+            }
+        }
+    }
+}
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:89]: All text files should end with a final new-line.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    public void run(String command, String[] args, Context context, PrintStream out) throws DDRInteractiveCommandException {
+        if (args.length > 2) {
+            out.println("showflags expects 2 arguments or less");
+            return;
+        }
+
+        StructureDescriptor currentStruct = null;
+        ConstantDescriptor currentFlag = null;
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:49]: Move declarations into the smallest possible scope.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        //Output all the constants of the flags structure 
+        if (args.length > 0){
+            for (StructureDescriptor i : currentStructures){
+                if (i.getName().equals(args[0])){
+                    currentStruct = i;
+                    if (args.length == 2){
+                        for (ConstantDescriptor j : i.getConstants()){ 
+                            if (j.getName().equals(args[1])){
+                                currentFlag = j;
+                                out.println(j.getName() + ": " + j.getValue());
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java:71]: `ConstantDescriptor.toString()` yields reasonable content: please just use that (like line 79 below).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by keithc-ca** [comment]
In reviewing this, I discovered we already have the `!buildflags` command. It only considers the `J9BuildFlags` class: what other classes are of interest and motivate this addition?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java (hunk 5 above)
Comment (keithc-ca) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShowFlags.java]: Please use tabs for indentation.

**Comment by dusanboskovic** [comment]
Yes, you are in fact right. I was following the advice that you had suggested, but after this information, it seems like I can skip to the next steps and make adjustments accordingly.

**Pull request closed without merging**